sources/api/early-boot-config/src/provider/cdrom.rs

```diff
@@ -0,0 +1,212 @@
+    // In VMWare, user data is supplied to the host via an XML file.  Within
+    // the XML file, there is a `PropertySection` that contains `Property` elements
+    // with attributes.  User data is base64 encoded inside a `Property` element with
+    // the attribute "user-data".
+    // <Property key="user-data" value="1234abcd"/>
+    fn ovf_user_data<P: AsRef<Path>>(path: P) -> Result<String> {
+        let path = path.as_ref();
+        let file = File::open(path).context(error::InputFileRead { path })?;
+        let reader = BufReader::new(file);
+
+        // Deserialize the OVF file, dropping everything we don't care about
+        let ovf: Environment =
+            serde_xml_rs::from_reader(reader).context(error::XMLDeserialize { path })?;
+
+        // We have seen the keys in the `Property` section be "namespaced" like
+        // "oe:key" or "of:key". We haven't been able to determine why certain files are
+        // this way, or where it comes from.  However, the namespacing doesn't seem to
+        // mean anything or change the way that user data is contained in the files.
+        // `serde_xml_rs` effectively ignores these namespaces and returns "key" / "value":
+        // https://github.com/Rreverser/serde-xml-rs/issues/64#issuecomment=540448434
+        let mut base64_str = String::new();
+        let user_data_key = "user-data";
+        for property in ovf.property_section.properties {
+            if property.key == user_data_key {
+                base64_str = property.value;
+                break;
+            }
+        }
+
+        // Base64 decode the &str
+        let decoded_bytes = base64::decode(&base64_str).context(error::Base64Decode {
+            base64_string: base64_str.to_string(),
+        })?;
+
+        // Create a valid utf8 str
+        let decoded = std::str::from_utf8(&decoded_bytes).context(error::InvalidUTF8 {
+            base64_string: base64_str.to_string(),
+        })?;
+
+        Ok(decoded.to_string())
+    }
+}
+
+impl PlatformDataProvider for CdromDataProvider {
+    fn platform_data(&self) -> super::Result<Vec<SettingsJson>> {
+        let mut output = Vec::new();
+
+        match Self::user_data() {
+            Err(e) => return Err(e).map_err(Into::into),
+            Ok(None) => warn!("No user data found."),
+            Ok(Some(s)) => output.push(s),
+        }
+        Ok(output)
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=
+
+// Minimal expected structure for an OVF file with user data
+#[derive(Debug, Deserialize)]
+struct Environment {
+    #[serde(rename = "PropertySection", default)]
+    pub property_section: PropertySection,
+}
+
+#[derive(Default, Debug, Deserialize)]
+struct PropertySection {
+    #[serde(rename = "Property", default)]
+    pub properties: Vec<Property>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Property {
+    pub key: String,
+    pub value: String,
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=
+
+mod error {
```

Comment:
In the error pattern I've suggested above, you could

```rust
impl Into<ProviderError> for cdrom::error::Error {
    /// ...
}
```

Which would avoid `map_err` calls.